File: packages/lhc-py/src/lhc/messages/internal/outcome.py
"""Ported from packages/lhc/src/messages/internal/outcome.ts. Phase 1 skeleton.

Mechanical ToolOutcome stamping: the outcome on a tool-activity summary is a
pure function of the record — paired-result presence and its isError flag —
and nothing else.
"""

from __future__ import annotations

from dataclasses import dataclass

from ...shared_tech.derivation import ToolOutcome


@dataclass(frozen=True, slots=True)
class PairedResult:
    is_error: bool


def derive_tool_outcome(paired_result: PairedResult | None) -> ToolOutcome:
    if paired_result is None:
        return "unknown"
    return "failed" if paired_result.is_error else "succeeded"

File: packages/lhc-rs/src/messages/internal/outcome.rs
//! Ported from packages/lhc/src/messages/internal/outcome.ts.
//!
//! Mechanical ToolOutcome stamping: the outcome on a tool-activity summary is a
//! pure function of the record — paired-result presence and its isError flag —
//! and nothing else.

use crate::shared_tech::derivation::ToolOutcome;

/// TS `{ isError: boolean }` argument shape for [`derive_tool_outcome`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PairedResult {
    pub is_error: bool,
}

pub fn derive_tool_outcome(paired_result: Option<PairedResult>) -> ToolOutcome {
    match paired_result {
        None => ToolOutcome::Unknown,
        Some(PairedResult { is_error: true }) => ToolOutcome::Failed,
        Some(PairedResult { is_error: false }) => ToolOutcome::Succeeded,
    }
}
